src/providers/moduleProvider.ts

```diff
@@ -0,0 +1,161 @@
+    const icon = {
+      light: svg,
+      dark: svg,
+    };
+    return icon;
+  }
+}
+
+export class TerraformModuleProvider implements vscode.TreeDataProvider<TerraformModule> {
+  private _onDidChangeTreeData: vscode.EventEmitter<TerraformModule | undefined | null | void> =
+    new vscode.EventEmitter<TerraformModule | undefined | null | void>();
+  readonly onDidChangeTreeData: vscode.Event<TerraformModule | undefined | null | void> =
+    this._onDidChangeTreeData.event;
+
+  constructor(ctx: vscode.ExtensionContext, public handler: ClientHandler) {
+    ctx.subscriptions.push(
+      vscode.commands.registerCommand('terraform.modules.refreshList', () => this.refresh()),
+      vscode.commands.registerCommand('terraform.modules.documentation', (module: TerraformModule) => {
+        vscode.env.openExternal(vscode.Uri.parse(module.doclink));
+      }),
+      vscode.window.onDidChangeActiveTextEditor(async (event: vscode.TextEditor | undefined) => {
+        if (event && vscode.workspace.workspaceFolders[0]) {
+          this.refresh();
+        }
+      }),
+    );
+  }
+
+  refresh(): void {
+    this._onDidChangeTreeData.fire();
+  }
+
+  getTreeItem(element: TerraformModule): TerraformModule | Thenable<TerraformModule> {
+    return element;
+  }
+
+  getChildren(element?: TerraformModule): vscode.ProviderResult<TerraformModule[]> {
+    if (element) {
+      return Promise.resolve(element.children);
+    } else {
+      const m = this.getModules();
+      return Promise.resolve(m);
+    }
+  }
+
+  getCollapseState(type: string): vscode.TreeItemCollapsibleState {
+    switch (type) {
+      case 'tfregistry':
+        return vscode.TreeItemCollapsibleState.Collapsed;
+      case 'local':
+        return vscode.TreeItemCollapsibleState.None;
+      case 'github':
+        return vscode.TreeItemCollapsibleState.None;
+      default:
+        return vscode.TreeItemCollapsibleState.None;
+    }
+  }
+
+  async getModules(): Promise<TerraformModule[]> {
+    const activeEditor = vscode.window.activeTextEditor;
+    if (activeEditor === undefined) {
+      return Promise.resolve([]);
+    }
+
+    const document = activeEditor.document;
+    if (document === undefined) {
+      return Promise.resolve([]);
+    }
+
+    const editor = document.uri;
+    const documentURI = Utils.dirname(editor);
+    const handler = this.handler.getClient(editor);
+
+    return await handler.client.onReady().then(async () => {
+      const params: ExecuteCommandParams = {
+        command: `${handler.commandPrefix}.terraform-ls.terraform.modulelist`,
+        arguments: [`uri=${documentURI}`],
+      };
+
+      const response = await handler.client.sendRequest(ExecuteCommandRequest.type, params);
+      if (response == null) {
+        return Promise.resolve([]);
+      }
+
+      const list = response.modules.map((m) => {
+        let deps: [TerraformModule];
+        if (m.depmodules === null) {
```

Comment:
I guess this is more on the LS implementation side, but if this is an array then I think we should be either returning empty array or not returning the field at all in the response, rather than returning `null`.

Reply:
True. Some of this is my habit of defensive programming to avoid failure at runtime but in this case I think you are correct that an empty array is more LSP protocol compliant than null

Reply:
Modified the LS implementation to always return an array, even if it is empty, with an error if present. This way we can determine client side if we want to show an error or show and empty window.

There is still too much object building here for my liking, will try to figure out if we can get these mapped via the LSP api.

Reply:
Tried to get the LS to return an already hydrated object, but the `workspace/executeCommand` API does not support returning objects, so will have to hydrate them ourselves.

Modified method to build out object, even with nested - nested modules.

Reply:
What do you mean by a "hydrated object"? 🚰  🤔 

> the workspace/executeCommand API does not support returning objects

I think that an object should fall within the `any` constraint in the spec?

https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_executeCommand

<img width="563" alt="Screenshot 2021-09-23 at 14 03 22" src="https://user-images.githubusercontent.com/287584/134511721-78f4d40b-a620-4545-86d7-7421ea9354a2.png">

Either way it seems like you managed to avoid the `null`?

Reply:
The `workspace/executeCommand` API returns `any`, which we have to 'know' what properties to access client side. If we used a custom request handler in the LS and used the `RequestType0` request type, we can specify an `interface` that will serialize the response into a Typescript object for us. That way we get type checking and intellisense client side.

The term `hydrate` is lingo I used to refer to data deserialized to structured objects.

We can leave the implementation as is and consider whether to move to a custom request handler if we implement more features in the future like this